### PR TITLE
Add functionality to upload etcd snapshot to s3 

### PR DESCRIPTION
**Pull request by iwilltry42**
Hi there,

This PR adds the functionality to RKE to upload etcd snapshots directly to AWS S3.
Feature request this targets: https://github.com/rancher/rke/issues/616
It's used from RKE with the following new flags:

-  `--s3`: feature toggle
- `--s3-bucket`: where to upload it (can be `<bucket>/<sub-path>`)
- `--s3-credentials FILE`: path to AWS credentials file
- `--s3-config FILE`: path to AWS config file (for region)

By default, the credentials + config from `~/.aws/credentials` + `~/.aws/config` will be used respectively.

Link to related PR to RKE: https://github.com/rancher/rke/pull/968

### CODE REVIEW

File: Dockerfile.dapper [hunk 1]
```diff
@@ -12,7 +12,7 @@ ENV GOLANG_ARCH_amd64=amd64 GOLANG_ARCH_arm=armv6l GOLANG_ARCH=GOLANG_ARCH_${ARC
     GOPATH=/go PATH=/go/bin:/usr/local/go/bin:${PATH} SHELL=/bin/bash
 
 RUN wget -O - https://storage.googleapis.com/golang/go1.9.2.linux-${!GOLANG_ARCH}.tar.gz | tar -xzf - -C /usr/local && \
-    go get github.com/rancher/trash && go get github.com/golang/lint/golint
+    go get github.com/rancher/trash && go get golang.org/x/lint/golint
```
Comment (galal-hussein) [inline Dockerfile.dapper]: i think this part was already resolved, can you rebase

**Comment by galal-hussein** [comment]
I think the main reason from using s3 is to make a centralized part for not only backup but for restore as well which is more time consuming when using files, can you check if it is possible to also restore from a snapshot placed on s3

File: Dockerfile.dapper (hunk 1 above)
Comment (iwilltry42) [reply Dockerfile.dapper]: Sure

**Comment by alena1108** [comment]
Closing in favor of s3 feature @guangbochen is working on

**Pull request closed without merging**